src/cpp/session/modules/SessionEnvironment.R

```diff
@@ -366,60 +366,40 @@
    return ("")
 })
 
+.rs.addFunction("editor", function(name, file = "", title = file, ...)
+{
+   # if 'name' is missing, we're likely being invoked by
+   # 'utils::file.edit()', so just edit the requested file
+   if (missing(name))
+      return(.Call("rs_editFile", file, PACKAGE = "(embedding)"))
+   
+   # otherwise, we're more likely being invoked by 'edit()', which
+   # requires the parsed file to be valid R code -- so handle that
+   
+   # if no file has been supplied, generate one for the user
+   if (is.null(file) || !nzchar(file)) {
+      file <- tempfile("rstudio-scratch-", fileext = ".R")
+      on.exit(unlink(file), add = TRUE)
+   }
+   
+   # write deparsed R object to file
+   deparsed <- if (is.function(name))
+      .rs.deparseFunction(name, useSource = TRUE, asString = FALSE)
+   else
+      deparse(name)
+   writeLines(deparsed, con = file)
+   
+   # invoke edit
+   .Call("rs_editFile", file, PACKAGE = "(embedding)")
+   
+   # attempt to parse-eval the generated content
+   eval(parse(file), envir = globalenv())
+   
+})
 
-.rs.addFunction("registerFunctionEditor", function() {
-
-   # save default editor
-   defaultEditor <- getOption("editor")
-
-   # ensure we have a scratch file
-   scratchFile <- tempfile()
-   cat("", file = scratchFile)
-
-   options(editor = function(name, file, title) {
-
-      # handle missing 'name' field
-      if (missing(name))
-         name <- NULL
-      
-      # use internal editor for files and functions, otherwise
-      # delegate to the default editor
-      if (is.null(name) || is.function(name)) {
-
-         # if no name then use file
-         if (is.null(name)) {
-            if (!is.null(file) && nzchar(file))
-               targetFile <- file
-            else
-               targetFile <- scratchFile
-         }
-         # otherwise it's a function, write it to a file for editing
-         else {
-            functionSrc <- .rs.deparseFunction(name, TRUE, FALSE)
-            targetFile <- scratchFile
-            writeLines(functionSrc, targetFile)
-         }
-
-         # invoke the RStudio editor on the file
-         if (.Call("rs_editFile", targetFile)) {
-
-            # try to parse it back in
-            newFunc <- try(eval.parent(parse(targetFile)),
-                           silent = TRUE)
-            if (inherits(newFunc, "try-error")) {
-               stop(newFunc, "You can attempt to correct the error using ",
-                    title, " = edit()")
-            }
-
-            return(newFunc)
-         }
-         else {
-            stop("Error occurred while editing function '", name, "'")
-         }
-      }
-      else
-         edit(name, file, title, editor=defaultEditor)
-   })
+.rs.addFunction("registerFunctionEditor", function()
+{
+   options(editor = .rs.editor)
```

Comment:
It looks like there was previously a provision for editing things besides files and functions (in which we invoked the default editor instead of our own). Was that intentionally removed?

Reply:
That was intentional, but it's worth asking if we do want to make that change. Effectively, this means the following:

Before: `edit("hello")` would open a file for editing in `notepad.exe`, with the contents `"hello"`. (Or, whatever default editor the user has registered.)

After: `edit("hello")` opens a file for editing in a modal RStudio editor, with the contents `"hello"`.

Also worth noting that this is Windows-only. RGui does something similar where it registers its own internal editor so that `edit("hello")` opens in an RGui editor as opposed to some external editor.

Reply:
I think that's reasonable and symmetric with what we do elsewhere.